File: examples/nfft/simple_test_threads.c
/*
 * Copyright (c) 2002, 2015 Jens Keiner, Stefan Kunis, Daniel Potts
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 51
 * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

/* $Id: simple_test.c 3198 2009-05-27 14:16:50Z keiner $ */
#include "config.h"

#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>
#ifdef HAVE_COMPLEX_H
#include <complex.h>
#endif
#include <omp.h>

#include "nfft3.h"
#include "infft.h"
//#include <time.h>

int main(void)
{
  NFFT(plan) p;
  const int N = 1000000;
  const int M = 1000000;
  ticks t0, t1;
  double t;

  printf("nthreads = %d\n", NFFT(get_num_threads)());

  /* init */
  FFTW(init_threads)();
  NFFT(init_1d)(&p,N,M);

  /* pseudo random nodes */
  NFFT(vrand_shifted_unit_double)(p.x,p.M_total);

  /* precompute psi, that is, the entries of the matrix B */
  t0 = getticks();
  if(p.flags & PRE_ONE_PSI)
      NFFT(precompute_one_psi)(&p);
  t1 = getticks();
  t = NFFT(elapsed_seconds(t1,t0));
  fprintf(stderr,"precompute elapsed time: %.3f seconds\n",t);

  /* pseudo random Fourier coefficients */
  NFFT(vrand_unit_complex)(p.f_hat,p.N_total);

  /* transformation */
  t0 = getticks();
  NFFT(trafo)(&p);
  t1 = getticks();
  t = NFFT(elapsed_seconds)(t1,t0);
  fprintf(stderr,"compute    elapsed time: %.3f seconds\n",t);
  fflush(stderr);
//  NFFT(vpr_complex)(p.f,p.M_total,"ndft, vector f");

  /* cleanup */
  NFFT(finalize)(&p);
  FFTW(cleanup_threads)();

  return EXIT_SUCCESS;
}

File: matlab/nfft/simple_test_threads.m
%SIMPLE_TEST_THREADS Example program: Basic usage principles for the
%set_num_threads function
%
%   Copyright (c) 2002, 2017 Jens Keiner, Stefan Kunis, Daniel Potts

% Copyright (c) 2002, 2017 Jens Keiner, Stefan Kunis, Daniel Potts
%
% This program is free software; you can redistribute it and/or modify it under
% the terms of the GNU General Public License as published by the Free Software
% Foundation; either version 2 of the License, or (at your option) any later
% version.
%
% This program is distributed in the hope that it will be useful, but WITHOUT
% ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
% FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
% details.
%
% You should have received a copy of the GNU General Public License along with
% this program; if not, write to the Free Software Foundation, Inc., 51
% Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
clear all

nthreads = nfft_get_num_threads();

% degree (bandwidth)
N = 32;

% number of nodes
M = 10000;

% nodes
x=rand(3,M)-0.5;

% Fourier coefficients
f_hat = rand(N^3,1)+1i*rand(N^3,1);

for threads = 1:nthreads
  nfft_set_num_threads(threads);

  % Create plan.
  plan = nfft_init_3d(N,N,N,M);

  % Set nodes.
  nfft_set_x(plan,x);

  % node-dependent precomputation
  nfft_precompute_psi(plan);

  % Set Fourier coefficients.
  nfft_set_f_hat(plan,double(f_hat));

  % transform
  tic
  nfft_trafo(plan);
  t = toc;

  % function values
  f = nfft_get_f(plan);

  % finalize plan
  nfft_finalize(plan);

  fprintf('Threads:%2d,  Time: %1.2e\n',threads,t);
end
